jlm/mlir/backend/JlmToMlirConverter.cpp

```diff
@@ -0,0 +1,222 @@
+  }
+  else
+  {
+    std::string unkown = "Unimplemented structural node: " + node.operation().debug_string();
+    JLM_UNREACHABLE(unkown.c_str());
+  }
+}
+
+::mlir::Value
+JlmToMlirConverter::ConvertSimpleNode(const rvsdg::simple_node & node, ::mlir::Block & block)
+{
+  if (auto bitsOp = dynamic_cast<const rvsdg::bitconstant_op *>(&(node.operation())))
+  {
+    auto value = bitsOp->value();
+    auto constOp = Builder_->create<::mlir::arith::ConstantIntOp>(
+        Builder_->getUnknownLoc(),
+        value.to_uint(),
+        value.nbits());
+    block.push_back(constOp);
+
+    return constOp;
+  }
+  else
+  {
+    std::string unkown = "Unimplemented simple node: " + node.operation().debug_string();
+    JLM_UNREACHABLE(unkown.c_str());
+  }
+}
+
+::mlir::Value
+JlmToMlirConverter::ConvertLambda(const llvm::lambda::node & lambdaNode, ::mlir::Block & block)
+{
+  ::llvm::SmallVector<::mlir::Type> arguments;
+  for (size_t i = 0; i < lambdaNode.nfctarguments(); ++i)
+  {
+    arguments.push_back(ConvertType(lambdaNode.fctargument(i)->type()));
+  }
+  ::llvm::ArrayRef argumentsArray(arguments);
+
+  ::llvm::SmallVector<::mlir::Type> results;
+  for (size_t i = 0; i < lambdaNode.nfctresults(); ++i)
+  {
+    results.push_back(ConvertType(lambdaNode.fctresult(i)->type()));
+  }
+  ::llvm::ArrayRef resultsArray(results);
+
+  ::llvm::SmallVector<::mlir::Type> lambdaRef;
+  auto refType = Builder_->getType<::mlir::rvsdg::LambdaRefType>(argumentsArray, resultsArray);
+  lambdaRef.push_back(refType);
+
+  ::llvm::SmallVector<::mlir::Value> inputs;
+  // TODO
+  // Populate the inputs
+
+  // Add function attributes, e.g., the function name and linkage
+  ::llvm::SmallVector<::mlir::NamedAttribute> attributes;
+  auto symbolName = Builder_->getNamedAttr(
+      Builder_->getStringAttr("sym_name"),
+      Builder_->getStringAttr(lambdaNode.name()));
+  attributes.push_back(symbolName);
+  auto linkage = Builder_->getNamedAttr(
+      Builder_->getStringAttr("linkage"),
+      Builder_->getStringAttr(jlm::llvm::linkage_to_string(lambdaNode.linkage())));
+  attributes.push_back(linkage);
+  ::llvm::ArrayRef<::mlir::NamedAttribute> attributesRef(attributes);
+
+  auto lambda = Builder_->create<::mlir::rvsdg::LambdaNode>(
+      Builder_->getUnknownLoc(),
+      lambdaRef,
+      inputs,
+      attributesRef);
+  block.push_back(lambda);
+
+  auto & lambdaBlock = lambda.getRegion().emplaceBlock();
+  auto regionResults = ConvertRegion(*lambdaNode.subregion(), lambdaBlock);
+  auto lambdaResult =
+      Builder_->create<::mlir::rvsdg::LambdaResult>(Builder_->getUnknownLoc(), regionResults);
+  lambdaBlock.push_back(lambdaResult);
+
+  return lambda;
+}
+
+::mlir::Type
+JlmToMlirConverter::ConvertType(const rvsdg::type & type)
+{
+  if (auto bt = dynamic_cast<const rvsdg::bittype *>(&type))
+  {
+    return Builder_->getIntegerType(bt->nbits());
+  }
+  else if (dynamic_cast<const llvm::loopstatetype *>(&type))
```

Comment:
`else if (is<const llvm::loopstatetype>(&type))`

Reply:
Changed